lib/auth/version.go

```diff
@@ -0,0 +1,89 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package auth
+
+import (
+	"context"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/auth/storage"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const (
+	// majorVersionConstraint is the major version constraint when previous major version must be
+	// present in the storage, if not - we must refuse to start.
+	// TODO(vapopov): DELETE IN 18.0.0
+	majorVersionConstraint = 18
+)
+
+// validateAndUpdateTeleportVersion validates that the major version persistent in the backend
+// meets our upgrade compatibility guide.
+func validateAndUpdateTeleportVersion(
```

Comment:
Should we have some kind of a hidden escape hatch? Like an (undocumented) environment variable that skips this check. Would be useful for development purposes.

Reply:
I might add something similar as we already have in `InitConfig`:
```
	// SkipPeriodicOperations turns off periodic operations
	// used in tests that don't need periodic operations.
	SkipPeriodicOperations bool
```
so this value going to be accessible only from code

Reply:
@vapopov Having it in the code only won't help much for development IMO. The idea is that when working on multiple different versions of Teleport you might need to switch back-and-forth between them on the same cluster, e.g. from v14 to v16 and vice versa to test something. That's been my experience before at least - and even though for production scenarios it is not supported upgrade/downgrade path, for development it works fine most of the time. So the idea is that there should be something devs can set in their environment to bypass this check.